File: golang/self_dividing_numbers.go
package main

import (
	"fmt"
	"strconv"
)

func selfDividingNumbers(left int, right int) []int {
	result := []int{}
	for i := left; i <= right; i++ {
		strNum := strconv.Itoa(i)
		isSelfDividing := true

		for _, char := range strNum {
			if char == '0' {
				isSelfDividing = false
				break
			}
			digit, _ := strconv.Atoi(string(char))
			if i%digit != 0 {
				isSelfDividing = false
				break
			}
		}

		if isSelfDividing {
			result = append(result, i)
		}
	}
	return result
}

func main() {
	fmt.Println(selfDividingNumbers(1, 22))
}

File: typescript/core/medium/self_dividing_numbers.ts
function selfDividingNumbers(left: number, right: number): number[] {
    const result: number[] = []
    for (let i = left; i <= right; i++) {
        if (i.toString().length > 1) {
            const stringRepresentation: string = i.toString()
            if (stringRepresentation.includes("0")) {
                continue
            }
            let isSelfDividingNumber: boolean = true
            for (let j = 0; j < stringRepresentation.length; j++) {
                const digit: number = +stringRepresentation[j]
                if (i % digit !== 0) {
                    isSelfDividingNumber = false
                    break
                }
            }
            if (isSelfDividingNumber) result.push(i)
        } else if (i.toString().length === 1) {
            result.push(i)
        }
    }
    return result
}

console.log(selfDividingNumbers(1, 22))
